Project/Assets/_Scripts/_new_Scripts/Weapons/RaycastGunViewModel.cs

```diff
@@ -2,123 +2,141 @@
+    {
+        private RaycastGunModel model;
 
-    public RaycastGunDefault Default;
-    public Transform ProjectileSpawn;
-    public GameObject LineRendererPrefab;
+        public RaycastGunDefault Default;
+        public Transform ProjectileSpawn;
+        public GameObject LineRendererPrefab;
 
-    private Vector3 rayDirection;
-    private LineRenderer activeRay;
+        private Vector3 rayDirection;
+        private LineRenderer activeRay;
 
-    private void Awake()
-    {
-        model = Default.LoadFromDefaults();
-    }
+        private void Awake()
+        {
+            model = Default.LoadFromDefaults();
+        }
 
-    public override void Attack()
-    {
-        model.TriggerTimer.Value += Time.deltaTime;
-        if (Time.time > model.FireTimer.Value && model.BulletsInMagazine.Value >= 1 && model.Reloading.Value == false && model.TriggerTimer.Value >= model.TimeToFire.Value)
+        public override void Attack()
         {
-            model.FireTimer.Value = Time.time + model.FireRate.Value;
+            model.ShotChargeTimer += Time.deltaTime;
+            if (CanAttack())
+            {
+                model.FireTimer = Time.time + model.FireRate;
 
-            ExecuteAttack();
+                ExecuteAttack();
+            }
         }
-    }
 
-    public override void StopAttack()
-    {
-        model.TriggerTimer.Value = 0f;
-    }
+        private bool CanAttack()
+        {
+            // There are bullets in magazine and player is not reloading already
+            if (model.BulletsInMagazine >= 1 && model.Reloading == false)
+            {
+                // Not waiting for shot to charge and firerate timer
+                if (model.ShotChargeTimer >= model.ShotChargeTime && Time.time > model.FireTimer)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
 
-    /// <summary>
-    /// Function called on pressing the ReloadButton
-    /// </summary>
-    public override void Reload()
-    {
-        if (model.BulletsInMagazine.Value < model.MagazineSize.Value && model.Reloading.Value == false)
+        public override void StopAttack()
         {
-            StartCoroutine(ExecuteReload());
+            model.ShotChargeTimer = 0f;
         }
-    }
 
-    /// <summary>
-    /// For now Instantiates ray for specified time and destroys it after
-    /// </summary>
-    private IEnumerator ShotEffect()
-    {
-        activeRay = Instantiate(LineRendererPrefab).GetComponent<LineRenderer>();
-        LineRenderer newRay = activeRay;
-        yield return new WaitForSeconds(model.ShotDuration.Value);
-        Destroy(newRay.gameObject);
-    }
+        /// <summary>
+        /// Function called on pressing the ReloadButton
+        /// </summary>
+        public override void Reload()
+        {
+            if (model.BulletsInMagazine < model.MagazineSize && model.Reloading == false)
```

Comment:
Tutaj bym tego ifa zmienił w jakąś prywatną funkcję tak jak CanAttack

Reply:
Właściwie ta funkcja to już jest coś w stylu CanAttack, bo sprawdza czy można przeładować i jak tak to uruchamia funkcję ExecuteReload. Wydaje mi się przesadą, bo cała funkcja to if i wywołanie innej funkcji, więc to samo co zrobiłoby CanReload.

Reply:
Ale tak właśnie zrobiłeś z funckcją `Attack()`

Reply:
https://pastebin.com/eJdp45EK
Myślę, że takie podwojenie ilości kodu pogarsza czytelność. Gdyby if był bardziej skomplikowany jak w Attack(), to jak najbardziej

Reply:
Chyba tak byłoby lepiej. I to ma sens, wtedy kod się przyjemnie czyta i wiadomo co się musi zadziać żeby gracz mógł zaatakować. To nie jest podwajanie ilości kodu, ponieważ mamy inny przypadek użycia tzn. gracz chce strzelić vs gracz chce przeładować.
```csharp
private bool CanReload()
{
           return model.BulletsInMagazine < model.MagazineSize && model.Reloading == false;
}
```

Reply:
No dobra, jak po zobaczeniu nadal stwierdzasz, że czytelniejsze, to pewnie masz rację